storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java

```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
+        }
+
+        LOG.info("The ordering result is {}", execsScheduled);
+
+        return execsScheduled;
+    }
 
-                for (Component neighborComp : sortedNeighbors) {
-                    Queue<ExecutorDetails> neighborCompExesToSched =
-                        compToExecsToSchedule.get(neighborComp.getId());
-                    if (!neighborCompExesToSched.isEmpty()) {
-                        execsScheduled.add(neighborCompExesToSched.poll());
-                        flag = true;
+    private List<ExecutorDetails> takeExecutors(Component currComp, int numExecs,
+                                                final Map<String, Component> componentMap,
+                                                final Map<String, Queue<ExecutorDetails>> compToExecsToSchedule) {
+        List<ExecutorDetails> execsScheduled = new ArrayList<>();
+        Queue<ExecutorDetails> currQueue = compToExecsToSchedule.get((currComp.getId()));
+        Set<String> sortedChildren = getSortedChildren(currComp, componentMap);
+
+        execsScheduled.add(currQueue.poll());
+
+        for (String childId: sortedChildren) {
+            Component childComponent = componentMap.get(childId);
+            Queue<ExecutorDetails> childQueue = compToExecsToSchedule.get(childId);
+            int childNumExecs = childQueue.size();
+            if (childNumExecs == 0) {
+                continue;
+            }
+            int numExecsToTake = 1;
+            if (isShuffleFromParentToChild(currComp, childComponent)) {
+                // if it's shuffle grouping, truncate
+                numExecsToTake = Math.max(1, childNumExecs / numExecs);
+            } // otherwise, one-by-one
+
+            for (int i = 0; i < numExecsToTake; i++) {
+                execsScheduled.addAll(takeExecutors(childComponent, childNumExecs, componentMap, compToExecsToSchedule));
+            }
+        }
+
+        return execsScheduled;
+    }
+
+    private Set<String> getSortedChildren(Component component, final Map<String, Component> componentMap) {
+        Set<String> children = component.getChildren();
+        Set<String> sortedChildren =
+                new TreeSet<String>((o1, o2) -> {
+                    Component child1 = componentMap.get(o1);
+                    Component child2 = componentMap.get(o2);
+                    boolean child1IsShuffle = isShuffleFromParentToChild(component, child1);
+                    boolean child2IsShuffle = isShuffleFromParentToChild(component, child2);
+
+                    if (child1IsShuffle && child2IsShuffle) {
+                        return o1.compareTo(o2);
+                    } else if (child1IsShuffle) {
+                        return 1;
+                    } else {
+                        return -1;
+                    }
+                });
+        sortedChildren.addAll(children);
+        return sortedChildren;
+    }
+
+    private boolean isShuffleFromParentToChild(Component parent, Component child) {
```

Comment:
Nit: could we rename this from `isShuffleFromParentToChild` to something more like `hasLocalityAwareGroupingFromParentToChild`?  I know it is longer, but in the future we may want to offer a way to expand this to more than just shuffle.